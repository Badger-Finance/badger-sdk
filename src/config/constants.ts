import { ARB_CONFIG } from './network/arbitrum.config';
import { AVAX_CONFIG } from './network/avalanche.config';
import { BSC_CONFIG } from './network/binance.config';
import { ETH_CONFIG } from './network/ethereum.config';
import { FTM_CONFIG } from './network/fantom.config';
import { LOCAL_CONFIG } from './network/local.config';
import { MATIC_CONFIG } from './network/polygon.config';

export const ONE_MIN_MS = 60 * 1000;
export const ONE_HOUR_MS = 60 * ONE_MIN_MS;
export const ONE_DAY_MS = 24 * ONE_HOUR_MS;
export const ONE_YEAR_MS = 365 * ONE_DAY_MS;

export const SUPPORTED_NETWORKS = [
  ETH_CONFIG,
  LOCAL_CONFIG,
  BSC_CONFIG,
  MATIC_CONFIG,
  ARB_CONFIG,
  AVAX_CONFIG,
  FTM_CONFIG,
];

export const VERSION = '1.0.77';
