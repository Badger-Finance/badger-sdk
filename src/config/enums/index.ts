export { Network } from './network.enum';
