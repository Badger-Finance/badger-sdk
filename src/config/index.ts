export * from './enums';
export * from './types';
export * from './constants';
export * from './interfaces';
export * from './network';
