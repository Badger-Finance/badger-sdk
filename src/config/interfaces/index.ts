export { NetworkConfig } from './network-config.interface';
