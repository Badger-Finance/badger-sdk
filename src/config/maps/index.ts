export * from './network.id.map';
