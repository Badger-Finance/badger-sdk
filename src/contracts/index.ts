/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
export type { BadgerTree } from "./BadgerTree";
export type { Digg } from "./Digg";
export type { Erc20 } from "./Erc20";
export type { Ibbtc } from "./Ibbtc";
export type { Registry } from "./Registry";
export type { RewardsLogger } from "./RewardsLogger";
export type { Sett } from "./Sett";
export type { StrategyV15 } from "./StrategyV15";
export type { VaultV15 } from "./VaultV15";

export { BadgerTree__factory } from "./factories/BadgerTree__factory";
export { Digg__factory } from "./factories/Digg__factory";
export { Erc20__factory } from "./factories/Erc20__factory";
export { Ibbtc__factory } from "./factories/Ibbtc__factory";
export { Registry__factory } from "./factories/Registry__factory";
export { RewardsLogger__factory } from "./factories/RewardsLogger__factory";
export { Sett__factory } from "./factories/Sett__factory";
export { StrategyV15__factory } from "./factories/StrategyV15__factory";
export { VaultV15__factory } from "./factories/VaultV15__factory";
