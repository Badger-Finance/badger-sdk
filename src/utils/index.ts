export * from './chunk-query-filter';
export * from './deployed-at.util';
export * from './key-by';
