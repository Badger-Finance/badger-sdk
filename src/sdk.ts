import { ethers } from 'ethers';
import { providers } from '@0xsequence/multicall';

import { Signer } from '@ethersproject/abstract-signer';
import { Networkish } from '@ethersproject/providers';

import { APIOptions, BadgerAPI } from './api';
import { BadgerGraph } from './graphql';

import { NetworkConfig } from './config/network/network.config';

import { DiggService } from './digg/digg.service';
import { ibBTCService } from './ibbtc/ibbtc.service';
import { RegistryService } from './registry/registry.service';
import { RewardsService } from './rewards/rewards.service';
import { VaultsService } from './vaults/vaults.service';
import { TokensService } from './tokens/tokens.service';

import { SDKProvider } from './config/types/sdk-provider';

import { SUPPORTED_NETWORKS } from './config/constants';

export interface SDKOptions extends APIOptions {
  provider: SDKProvider | string;
}

export class BadgerSDK {
  private static initialized = false;
  private loading: Promise<void>;

  public address?: string;
  public api: BadgerAPI;
  public graph: BadgerGraph;
  public config: NetworkConfig;
  public provider: providers.MulticallProvider;
  public signer?: Signer;

  readonly registry: RegistryService;
  readonly digg: DiggService;
  readonly ibbtc: ibBTCService;
  readonly rewards: RewardsService;
  readonly tokens: TokensService;
  readonly vaults: VaultsService;

  constructor({ network, provider, baseURL }: SDKOptions) {
    if (!BadgerSDK.initialized) {
      for (const config of SUPPORTED_NETWORKS) {
        NetworkConfig.register(config);
      }
      BadgerSDK.initialized = true;
    }

    const sdkProvider = BadgerSDK.getSdkProvider(provider);

    this.provider = new providers.MulticallProvider(sdkProvider);
    this.signer = sdkProvider.getSigner();
    this.loading = this.initialize();
    this.config = NetworkConfig.getConfig(network);
    this.api = new BadgerAPI({
      network: this.config.network,
      baseURL,
    });
    this.graph = new BadgerGraph({
      network: this.config.network,
    });
    this.digg = new DiggService(this);
    this.ibbtc = new ibBTCService(this);
    this.registry = new RegistryService(this);
    this.rewards = new RewardsService(this);
    this.tokens = new TokensService(this);
    this.vaults = new VaultsService(this);
  }

  ready() {
    return Promise.all([this.loading, this.registry.ready(), this.rewards.ready()]);
  }

  update(network: Networkish, provider: SDKProvider) {
    this.updateNetwork(network);
    this.updateProvider(provider);
  }

  updateNetwork(network: Networkish) {
    this.config = NetworkConfig.getConfig(network);
    this.api = new BadgerAPI({
      network: this.config.network,
      baseURL: this.api.baseURL,
    });
    this.graph = new BadgerGraph({
      network: this.config.network,
    });
  }

  updateProvider(provider: SDKProvider) {
    const sdkProvider = BadgerSDK.getSdkProvider(provider);

    this.provider = new providers.MulticallProvider(sdkProvider);
    this.signer = sdkProvider.getSigner();
  }

  private static getSdkProvider(provider: SDKProvider | string): SDKProvider {
    let sdkProvider: SDKProvider;

    if (typeof provider === 'string') {
      sdkProvider = new ethers.providers.JsonRpcBatchProvider(provider);
    } else {
      sdkProvider = provider;
    }

    return sdkProvider;
  }

  private async initialize() {
    try {
      if (this.signer) {
        this.address = await this.signer.getAddress();
      }
    } catch {} // ignore errors from getting address
  }
}
