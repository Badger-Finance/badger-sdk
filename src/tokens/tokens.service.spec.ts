import { BigNumber, ethers } from 'ethers';
import { MockProxy } from 'jest-mock-extended';
import { TransactionStatus } from '../config';
import { Erc20 } from '../contracts';
import { BadgerSDK } from '../sdk';
import { TEST_ADDR, TRANSACTION_SUCCESS } from '../tests/tests.constants';
import { mockSDK, mockToken } from '../tests/tests.utils';

describe('tokens.service', () => {
  let sdk: BadgerSDK;
  let token: MockProxy<Erc20>;

  beforeEach(async () => {
    sdk = mockSDK();
    token = mockToken();

    await sdk.ready();
  });

  describe('revoke', () => {
    it('reports failure with no address for the sdk', async () => {
      sdk.address = undefined;
      const result = await sdk.tokens.revoke({
        token: TEST_ADDR,
        spender: TEST_ADDR,
      });
      expect(result).toMatch(TransactionStatus.Failure);
    });

    it('reports success with given a proper signer', async () => {
      jest
        .spyOn(token, 'increaseAllowance')
        .mockImplementation(async (_s, _a, _o) => TRANSACTION_SUCCESS);
      const result = await sdk.tokens.revoke({
        token: TEST_ADDR,
        spender: TEST_ADDR,
      });
      expect(result).toMatch(TransactionStatus.Success);
    });
  });

  describe('verifyOrIncreaseAllowance', () => {
    it('reports failure with no address for the sdk', async () => {
      sdk.address = undefined;
      const result = await sdk.tokens.verifyOrIncreaseAllowance({
        token: TEST_ADDR,
        spender: TEST_ADDR,
        amount: BigNumber.from('15000000000000000000'), // 15
      });
      expect(result).toMatch(TransactionStatus.Failure);
    });

    it('reports success with no action if a spender has proper allowance', async () => {
      jest
        .spyOn(token, 'allowance')
        .mockImplementation(async (_a, _t) =>
          BigNumber.from('15000000000000000000'),
        );
      const result = await sdk.tokens.verifyOrIncreaseAllowance({
        token: TEST_ADDR,
        spender: TEST_ADDR,
        amount: BigNumber.from('15000000000000000000'), // 15
      });
      expect(result).toMatch(TransactionStatus.Success);
    });

    it('reports success with after increasing allowance if a spender does not have proper allowance', async () => {
      let checkedPrompt = false;
      let checkedSigned = false;
      let checkedSuccess = false;
      const onApprovePrompt = () => {
        checkedPrompt = true;
      };
      const onApproveSigned = () => {
        checkedSigned = true;
      };
      const onApproveSuccess = () => {
        checkedSuccess = true;
      };
      jest
        .spyOn(token, 'allowance')
        .mockImplementation(async (_a, _t) => BigNumber.from('0'));
      jest
        .spyOn(token, 'increaseAllowance')
        .mockImplementation(async (_s, _a, _o) => TRANSACTION_SUCCESS);
      const result = await sdk.tokens.verifyOrIncreaseAllowance({
        token: TEST_ADDR,
        spender: TEST_ADDR,
        amount: BigNumber.from('15000000000000000000'), // 15
        onApprovePrompt,
        onApproveSigned,
        onApproveSuccess,
      });
      expect(result).toMatch(TransactionStatus.Success);
      expect(checkedPrompt).toBeTruthy();
      expect(checkedSigned).toBeTruthy();
      expect(checkedSuccess).toBeTruthy();

      // set up a failure if allowance is called - let's call the cache
      jest.spyOn(token, 'allowance').mockImplementation(async (_a, _t) => {
        throw new Error('Test Error: failed caching if thrown');
      });
      const secondResult = await sdk.tokens.verifyOrIncreaseAllowance({
        token: TEST_ADDR,
        spender: TEST_ADDR,
        amount: BigNumber.from('15000000000000000000'), // 15
        onApprovePrompt,
        onApproveSigned,
        onApproveSuccess,
      });
      expect(secondResult).toMatch(TransactionStatus.Success);
    });

    it('reports canceled if user does not sign the transaction', async () => {
      jest
        .spyOn(token, 'allowance')
        .mockImplementation(async (_a, _t) => BigNumber.from('0'));
      jest
        .spyOn(token, 'increaseAllowance')
        .mockImplementation(async (_s, _a, _o) => {
          throw new Error(
            'Expected test error: simulating cancelling transaction',
          );
        });
      const canceledIncrease = await sdk.tokens.verifyOrIncreaseAllowance({
        token: TEST_ADDR,
        spender: TEST_ADDR,
        amount: BigNumber.from('15000000000000000000'), // 15
      });
      expect(canceledIncrease).toMatch(TransactionStatus.Canceled);
    });

    it('reports canceled if user does not sign the transaction', async () => {
      jest
        .spyOn(token, 'allowance')
        .mockImplementation(async (_a, _t) => BigNumber.from('0'));
      jest
        .spyOn(token, 'increaseAllowance')
        .mockImplementation(async (_s, _a, _o) => {
          throw new Error(
            'Expected test error: simulating cancelling transaction',
          );
        });
      let checkedRejected = false;
      const onRejection = () => {
        checkedRejected = true;
      };
      const canceledIncrease = await sdk.tokens.verifyOrIncreaseAllowance({
        token: TEST_ADDR,
        spender: TEST_ADDR,
        amount: BigNumber.from('15000000000000000000'), // 15
        onRejection,
      });
      expect(canceledIncrease).toMatch(TransactionStatus.Canceled);
      expect(checkedRejected).toBeTruthy();
    });

    it('reports failure if increase allowance call reverts or fails', async () => {
      jest
        .spyOn(token, 'allowance')
        .mockImplementation(async (_a, _t) => BigNumber.from('0'));
      jest
        .spyOn(token, 'increaseAllowance')
        .mockImplementation(async (_s, _a, _o) => {
          return {
            chainId: 1,
            gasLimit: BigNumber.from('1000000'),
            nonce: 1,
            data: '',
            value: ethers.constants.Zero,
            hash: TEST_ADDR,
            from: TEST_ADDR,
            confirmations: 0,
            wait: async (): Promise<ethers.ContractReceipt> => {
              throw new Error(
                'Expected test error: simulating cancelling transaction',
              );
            },
          };
        });
      let checkedError = false;
      const onError = () => {
        checkedError = true;
      };
      const canceledIncrease = await sdk.tokens.verifyOrIncreaseAllowance({
        token: TEST_ADDR,
        spender: TEST_ADDR,
        amount: BigNumber.from('15000000000000000000'), // 15
        onError,
      });
      expect(canceledIncrease).toMatch(TransactionStatus.Failure);
      expect(checkedError).toBeTruthy();
    });
  });

  describe('loadToken', () => {
    it('returns requested token information', async () => {
      const loadedToken = await sdk.tokens.loadToken(TEST_ADDR);
      expect(loadedToken).toMatchSnapshot();

      // verify token information caching
      jest.spyOn(token, 'name').mockImplementation(async () => {
        throw new Error('Test Error: failed caching if thrown');
      });
      const duplicateToken = await sdk.tokens.loadToken(TEST_ADDR);
      expect(duplicateToken).toMatchObject(loadedToken);
    });
  });

  describe('loadTokens', () => {
    const testAddressTwo = '0x1622bF67e6e5747b81866fE0b85178a93C7F86e3';
    const testAddressThree = '0xff628b747c4e70825af24e3d59748bac477dcbf6';

    it('returns requested token information', async () => {
      const loadedTokens = await sdk.tokens.loadTokens([
        TEST_ADDR,
        testAddressTwo,
      ]);
      expect(loadedTokens).toMatchSnapshot();

      let calls = 0;
      // verify token information caching
      jest.spyOn(token, 'name').mockImplementation(async () => {
        if (calls === 0) {
          return 'Success Test Name Token';
        }
        calls += 1;
        throw new Error('Test Error: failed caching if thrown');
      });
      const partialFailureTokens = await sdk.tokens.loadTokens([
        TEST_ADDR,
        testAddressTwo,
        testAddressThree,
      ]);
      expect(partialFailureTokens).toMatchSnapshot();
    });
  });
});
