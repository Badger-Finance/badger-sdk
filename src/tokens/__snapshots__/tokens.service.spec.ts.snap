// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokens.service loadAllowances returns requested allowance information 1`] = `
Object {
  "0x1622bF67e6e5747b81866fE0b85178a93C7F86e3": Object {
    "hex": "0xd02ab486cedc0000",
    "type": "BigNumber",
  },
  "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484": Object {
    "hex": "0xd02ab486cedc0000",
    "type": "BigNumber",
  },
}
`;

exports[`tokens.service loadBalances returns requested balance information 1`] = `
Object {
  "0x1622bF67e6e5747b81866fE0b85178a93C7F86e3": Object {
    "hex": "0xd02ab486cedc0000",
    "type": "BigNumber",
  },
  "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484": Object {
    "hex": "0xd02ab486cedc0000",
    "type": "BigNumber",
  },
}
`;

exports[`tokens.service loadToken returns requested token information 1`] = `
Object {
  "address": "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484",
  "decimals": 18,
  "name": "Test Token",
  "symbol": "JTT",
}
`;

exports[`tokens.service loadTokens returns requested token information 1`] = `
Object {
  "0x1622bF67e6e5747b81866fE0b85178a93C7F86e3": Object {
    "address": "0x1622bF67e6e5747b81866fE0b85178a93C7F86e3",
    "decimals": 18,
    "name": "Test Token",
    "symbol": "JTT",
  },
  "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484": Object {
    "address": "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484",
    "decimals": 18,
    "name": "Test Token",
    "symbol": "JTT",
  },
}
`;

exports[`tokens.service loadTokens returns requested token information 2`] = `
Object {
  "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484": Object {
    "address": "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484",
    "decimals": 18,
    "name": "Test Token",
    "symbol": "JTT",
  },
  "0xfF628B747C4E70825AF24e3d59748bac477dcbF6": Object {
    "address": "0xfF628B747C4E70825AF24e3d59748bac477dcbF6",
    "decimals": 18,
    "name": "",
    "symbol": "",
  },
}
`;
